htsget-search/src/htsget/cram_search.rs

```diff
@@ -0,0 +1,324 @@
+
+        let mut reader = File::open(cram_path)
+            .map(BufReader::new)
+            .map(noodles_cram::Reader::new)
+            .map_err(|_| HtsGetError::io_error("Reading CRAM"))?;
+
+        reader.read_file_definition().map_err(|_| HtsGetError::io_error("Reading CRAM file definition"))?;
+
+        let header = reader
+            .read_file_header()
+            .map_err(|_| HtsGetError::io_error("Reading CRAM header"))?
+            .parse()
+            .map_err(|_| HtsGetError::io_error("Parsing CRAM header"))?;
+
+        Ok((reader, header))
+    }
+
+    /// Get key for storage object.
+    fn get_keys_from_id(&self, id: &str) -> (String, String) {
+        let cram_key = format!("{}.cram", id);
+        let crai_key = format!("{}.crai", cram_key);
+        (cram_key, crai_key)
+    }
+
+    /// Returns mapped and placed unmapped ranges
+    fn get_byte_ranges_for_all_reads(
+        &self,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+    ) -> Result<Vec<BytesRange>> {
+        Self::get_bytes_ranges(
+            crai_index,
+            cram_reader,
+            |_| true,
+        )
+    }
+
+    /// Returns only unplaced unmapped ranges
+    fn get_byte_ranges_for_unmapped_reads(
+        &self,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+    ) -> Result<Vec<BytesRange>> {
+        Self::get_bytes_ranges(
+            crai_index,
+            cram_reader,
+            |record| {
+                record.reference_sequence_id().map_or_else(
+                    || true,
+                    |ref_seq| i32::from(ref_seq) == reference_sequence_id::UNMAPPED,
+                )
+            },
+        )
+    }
+
+    /// Returns reads for a given reference name and an optional sequence range
+    fn get_byte_ranges_for_reference_name(
+        &self,
+        reference_name: &str,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+        cram_header: Header,
+    ) -> Result<Vec<BytesRange>> {
+        let maybe_cram_ref_seq = cram_header.reference_sequences().get_full(reference_name);
+
+        let byte_ranges = match maybe_cram_ref_seq {
+            None => Err(HtsGetError::not_found(format!(
+                "Reference name not found: {}",
+                reference_name
+            ))),
+            Some((ref_seq_id, _, _)) => {
+                let cram_ref_seq_idx = ReferenceSequenceId::try_from(ref_seq_id as i32)
+                    .map_err(|_| HtsGetError::invalid_input("Invalid reference sequence id"))?;
+                Self::get_bytes_ranges(
+                    crai_index,
+                    cram_reader,
+                    |record| record.reference_sequence_id() == Some(cram_ref_seq_idx),
+                )
+            }
+        }?;
+        Ok(byte_ranges)
+    }
+
+    /// Get bytes ranges using the index.
+    fn get_bytes_ranges<F>(
```

Comment:
I think this should be `get_byte_ranges`, so it is consistent with the rest of function names